code/game/objects/items/devices/painter/decal_painter.dm

```diff
@@ -0,0 +1,115 @@
+/datum/painter/decal
+	module_name = "decal painter"
+	module_state = "decal_painter"
+	/// icon that contains the decal sprites
+	var/decal_icon = 'icons/turf/decals.dmi'
+	/// icon_state of the selected decal
+	var/decal_state = "warn_box"
+	var/decal_dir = SOUTH
+	/// When removal_mode is TRUE the decal painter will remove decals instead
+	var/removal_mode = FALSE
+	var/max_decals = 3
+	var/static/list/decal_blacklist = typecacheof(
+		list(
+			/obj/effect/turf_decal/raven,
+			/obj/effect/turf_decal/weather,
+			/obj/effect/turf_decal/stripes/asteroid,
+			/obj/effect/turf_decal/tile,
+			/obj/effect/turf_decal/sand
+		)
+	)
+	/// Assoc list with icon_state of the decal as the key, and decal path as the value.
+	var/static/list/lookup_cache_decals = list()
+
+/datum/painter/decal/New(obj/item/painter/parent_painter)
+	. = ..()
+	if(!length(lookup_cache_decals))
+		for(var/D in subtypesof(/obj/effect/turf_decal))
+			var/obj/effect/turf_decal/decal = D
+			if(decal in decal_blacklist)
+				continue
+			lookup_cache_decals[decal::icon_state] = decal
+
+/datum/painter/decal/paint_atom(atom/target, mob/user)
+	if(!istype(target, /turf/simulated/floor))
+		to_chat(user, "<span class='warning'>[holder] can only be used on flooring.</span>")
+		return FALSE
+	var/decals = target.GetComponents(/datum/component/decal)
```

Comment:
Dunno if there's a way that this might be doable without calling `GetComponents`, such as with signals? Also don't know if we'd want signals on every single decal ever, though.
```suggestion
	var/list/decals = target.GetComponents(/datum/component/decal)
```